### PR TITLE
Implement `canonize` for canonical form transformation of `MatrixProductState`s

### LINKED ISSUE #45: Canonical form for `MatrixProductState` and `MatrixProductOperator`
### Summary 
Add a canonical form transformation for a given `MatrixProductState`. This will be useful for different algorithms.

### PR DESCRIPTION
**Pull request by jofrevalles**
### Summary
In this PR we add the function `canonize` (fix #45), which takes a `MatrixProductState{Open}` and it transforms it to the desired canonical form. This function accepts a `center`, which determines the center of orthogonality, which can be an `Integer` or `UnitRange`. It performs `svd` to ensure that the tensors on the left and right of the `center` are left and right canonical, respectively. It also can limit the bond dimension `chi` to the desired value for the tensors outside of the `center`. We also added tests to ensure that it works accordingly.
 
### Example
```julia
julia> using Tenet
julia> ψ = rand(MatrixProductState{Open}, 16, 2, 8)
TensorNetwork{MatrixProductState{Open}}(#tensors=16, #inds=31)
julia> ϕ = canonize(ψ, 8)
TensorNetwork{MatrixProductState{Open}}(#tensors=16, #inds=31)
julia> A = tensors(ϕ, 4); ein"ijk,ilk->jl"(A, conj(A))
16×16 Matrix{Float64}:
  1.0           2.00021e-16   1.55642e-16  -3.6175e-17    7.2592e-17    1.22594e-17   4.15944e-16  …  -1.33193e-16  -2.02137e-16   2.47257e-16  -9.46201e-17   1.3618e-16    6.85258e-17
  2.00021e-16   1.0           1.1956e-16    1.23764e-16   4.2185e-16   -4.7835e-16   -2.90511e-16     -1.60436e-17  -5.19301e-16   1.33808e-16  -1.58435e-16   9.47992e-17  -2.75936e-16
  1.55642e-16   1.1956e-16    1.0          -1.44702e-16  -1.55503e-17  -3.18376e-17   3.74575e-17      3.60339e-17  -1.89068e-16  -2.46219e-16  -7.9332e-17   -3.10887e-16   1.67774e-17
 -3.6175e-17    1.23764e-16  -1.44702e-16   1.0          -5.88042e-17   2.82227e-16  -7.15532e-17     -5.30028e-18  -8.86827e-17  -1.06777e-16  -8.00802e-17   7.91298e-17  -3.02182e-16
  7.2592e-17    4.2185e-16   -1.55503e-17  -5.88042e-17   1.0          -3.98925e-17  -1.44373e-16     -1.80332e-16  -2.1445e-16   -1.43922e-16  -1.88635e-16  -3.0402e-17   -2.28285e-16
  1.22594e-17  -4.7835e-16   -3.18376e-17   2.82227e-16  -3.98925e-17   1.0          -7.3377e-17   …   2.54161e-16   2.54636e-16  -1.84121e-17  -3.17101e-17   4.80834e-17   7.39615e-17
  4.15944e-16  -2.90511e-16   3.74575e-17  -7.15532e-17  -1.44373e-16  -7.3377e-17    1.0             -2.66094e-16  -4.13278e-17  -2.43669e-17   2.0519e-16   -3.93628e-16  -1.44632e-16
 -5.37008e-17  -4.56306e-17  -8.04112e-17  -4.56559e-16  -6.4265e-18   -3.35598e-16   6.23502e-16      2.66407e-17  -1.7158e-16   -1.41715e-18   6.19677e-17   6.93524e-18   2.79394e-16
  8.05436e-17   1.42453e-16   1.71197e-16   2.08164e-16  -5.35533e-17   1.02004e-16   6.79889e-16     -1.44901e-16  -5.18267e-17  -1.0725e-16    2.23824e-17  -3.37965e-16  -4.39888e-16
  1.61461e-16   1.08783e-16  -1.16969e-16   1.56489e-16   4.03115e-16   5.58961e-17   1.18552e-16     -1.39082e-17  -1.99405e-16   5.04873e-16  -9.1763e-17    3.91543e-17   3.13676e-17
 -1.33193e-16  -1.60436e-17   3.60339e-17  -5.30028e-18  -1.80332e-16   2.54161e-16  -2.66094e-16  …   1.0          -1.15568e-17  -1.47709e-16   1.73387e-16   1.80294e-16   1.60105e-16
 -2.02137e-16  -5.19301e-16  -1.89068e-16  -8.86827e-17  -2.1445e-16    2.54636e-16  -4.13278e-17     -1.15568e-17   1.0          -3.229e-16     4.53753e-16  -1.37335e-16  -6.98021e-17
  2.47257e-16   1.33808e-16  -2.46219e-16  -1.06777e-16  -1.43922e-16  -1.84121e-17  -2.43669e-17     -1.47709e-16  -3.229e-16     1.0          -5.98558e-17   7.25681e-17   2.80239e-16
 -9.46201e-17  -1.58435e-16  -7.9332e-17   -8.00802e-17  -1.88635e-16  -3.17101e-17   2.0519e-16       1.73387e-16   4.53753e-16  -5.98558e-17   1.0           1.24428e-16   3.75688e-19
  1.3618e-16    9.47992e-17  -3.10887e-16   7.91298e-17  -3.0402e-17    4.80834e-17  -3.93628e-16      1.80294e-16  -1.37335e-16   7.25681e-17   1.24428e-16   1.0           1.25388e-16
  6.85258e-17  -2.75936e-16   1.67774e-17  -3.02182e-16  -2.28285e-16   7.39615e-17  -1.44632e-16  …   1.60105e-16  -6.98021e-17   2.80239e-16   3.75688e-19   1.25388e-16   1.0
julia> B = tensors(ϕ, 12); ein"ijk,ljk->il"(B, conj(B))
32×32 Matrix{Float64}:
  1.0          -1.01264e-16   3.29597e-17  -2.1684e-17    7.1991e-17   -1.74557e-16   4.16334e-17  …   2.00297e-32   1.30963e-32   0.0           5.20417e-17   4.85723e-17  -4.0766e-17
 -1.01264e-16   1.0          -7.48099e-17  -3.85976e-17   1.64799e-17   1.74475e-16   1.84314e-17      1.75741e-32   1.07852e-32   1.54074e-33  -1.73472e-16   8.67362e-19  -1.82146e-16
  3.29597e-17  -7.48099e-17   1.0          -6.09322e-17  -4.51028e-17  -6.1149e-17   -1.64799e-17      5.10371e-33  -1.77186e-32   0.0           2.68882e-17  -1.04083e-17   1.21431e-17
 -2.1684e-17   -3.85976e-17  -6.09322e-17   1.0           1.73472e-18  -7.08797e-17   2.0383e-17       2.48445e-32   6.16298e-33   8.85928e-33   1.88651e-17   3.46945e-18  -1.06252e-17
  7.1991e-17    1.64799e-17  -4.51028e-17   1.73472e-18   1.0           9.56266e-17  -1.73472e-17      4.62223e-33   0.0          -7.70372e-34   2.77556e-16   3.46945e-18   2.77556e-17
 -1.74557e-16   1.74475e-16  -6.1149e-17   -7.08797e-17   9.56266e-17   1.0           2.33429e-16  …   3.58896e-27  -6.39421e-28   3.2983e-27    2.28767e-17   2.81893e-17  -7.24789e-17
  4.16334e-17   1.84314e-17  -1.64799e-17   2.0383e-17   -1.73472e-17   2.33429e-16   1.0             -9.62965e-34   7.70372e-34  -9.24446e-33  -6.4642e-32    3.46945e-17  -6.07153e-18
  2.25514e-17  -7.02563e-17   7.63278e-17   6.245e-17     1.38778e-17   1.99493e-17  -3.46945e-18     -1.09778e-32   1.00148e-32  -7.70372e-33  -6.59195e-17   6.245e-17     4.16334e-17
  9.71445e-17  -9.45424e-17   1.14492e-16   5.72459e-17   3.46945e-17  -5.63785e-17   1.73472e-17      1.20366e-18  -8.55969e-19  -5.97008e-20   1.04083e-17  -1.38778e-17   1.04083e-16
  ⋮                                                                     ⋮                          ⋱                                                           ⋮            
 -3.15853e-32  -9.05187e-33  -1.20371e-32  -2.32556e-32   5.77779e-33  -5.42224e-27   1.56e-32         2.08167e-17   8.32667e-17  -2.77556e-17  -9.26063e-18   7.692e-18    -1.6449e-17
 -1.50223e-32  -3.87112e-32  -1.71408e-32  -2.07037e-33   3.46667e-33   1.05074e-26  -4.23705e-33     -1.21431e-16   5.55112e-17   9.71445e-17   8.94486e-17   4.24816e-17   1.18325e-16
 -2.50371e-33   6.64446e-33   4.55001e-33  -4.21297e-33  -4.62223e-33  -3.8008e-27   -9.82224e-33  …   2.68882e-17   1.07553e-16  -6.93889e-17  -2.65699e-17   1.55548e-16  -3.69e-17
  2.00297e-32   1.75741e-32   5.10371e-33   2.48445e-32   4.62223e-33   3.58896e-27  -9.62965e-34      1.0          -1.38778e-17  -4.16334e-17   2.47101e-17   4.15695e-17   4.25864e-17
  1.30963e-32   1.07852e-32  -1.77186e-32   6.16298e-33   0.0          -6.39421e-28   7.70372e-34     -1.38778e-17   1.0          -2.22045e-16  -1.75484e-17  -3.31585e-17  -2.74651e-17
  0.0           1.54074e-33   0.0           8.85928e-33  -7.70372e-34   3.2983e-27   -9.24446e-33     -4.16334e-17  -2.22045e-16   1.0           2.38417e-17   1.31197e-17   3.0879e-17
  5.20417e-17  -1.73472e-16   2.68882e-17   1.88651e-17   2.77556e-16   2.28767e-17  -6.4642e-32       2.47101e-17  -1.75484e-17   2.38417e-17   1.0           9.36751e-17   1.7911e-16
  4.85723e-17   8.67362e-19  -1.04083e-17   3.46945e-18   3.46945e-18   2.81893e-17   3.46945e-17  …   4.15695e-17  -3.31585e-17   1.31197e-17   9.36751e-17   1.0           1.80411e-16
 -4.0766e-17   -1.82146e-16   1.21431e-17  -1.06252e-17   2.77556e-17  -7.24789e-17  -6.07153e-18      4.25864e-17  -2.74651e-17   3.0879e-17    1.7911e-16    1.80411e-16   1.0
julia> [size(tensors(ϕ, i)) for i in 1:length(ϕ)]
16-element Vector{Tuple{Int64, Int64, Vararg{Int64}}}:
 (2, 2)
 (2, 4, 2)
 (4, 8, 2)
 (8, 16, 2)
 (16, 32, 2)
 (32, 32, 2)
 (32, 16, 2)
 (16, 8, 2)
 (8, 16, 2)
 (16, 32, 2)
 (32, 32, 2)
 (32, 16, 2)
 (16, 8, 2)
 (8, 4, 2)
 (4, 2, 2)
 (2, 2)

julia> ϕ2 = canonize(ψ, 8; chi = 4)
TensorNetwork{MatrixProductState{Open}}(#tensors=16, #inds=31)
julia> [size(tensors(ϕ2, i)) for i in 1:length(ϕ2)]
16-element Vector{Tuple{Int64, Int64, Vararg{Int64}}}:
 (2, 2)
 (2, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 8, 2)
 (8, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 4, 2)
 (4, 2, 2)
 (2, 2)
julia> A = tensors(ϕ2, 4); ein"ijk,ilk->jl"(A, conj(A))
4×4 Matrix{Float64}:
  1.0          -1.01412e-16  -5.00915e-16   2.35041e-16
 -1.01412e-16   1.0           1.14853e-17  -5.86129e-16
 -5.00915e-16   1.14853e-17   1.0          -3.02073e-16
  2.35041e-16  -5.86129e-16  -3.02073e-16   1.0
julia> B = tensors(ϕ2, 12); ein"ijk,ljk->il"(B, conj(B))
4×4 Matrix{Float64}:
  1.0          8.67362e-19  -6.245e-17    -1.04951e-16
  8.67362e-19  1.0           3.40006e-16   1.9082e-16
 -6.245e-17    3.40006e-16   1.0           2.63678e-16
 -1.04951e-16  1.9082e-16    2.63678e-16   1.0
```

### CODE REVIEW

**Comment by mofeing** [comment]
> I don't understand why I have this error in the test. Why is `using OMEinsum` not enough to load `@ein_str`? I am using this locally and I don't have issues.

Strange error. Maybe `OMEinsum` is not added to the test project but anyway, you should try to use `contract` and `svd` with `Tensor`s.

I'm gonna pull the code to fix it and refactor it a lil.

**Comment by jofrevalles** [comment]
> Strange error. Maybe `OMEinsum` is not added to the test project but anyway, you should try to use `contract` and `svd` with `Tensor`s.

Can I use `contract` with certain indices only?

> I'm gonna pull the code to fix it and refactor it a lil.

Okay, however you prefer. I can also do it myself.

**Comment by mofeing** [comment]
> Can I use `contract` with certain indices only?

Yes! That's the 3 argument. Pass a list or tuple of `Symbol`s.

**Comment by jofrevalles** [comment]
> > Can I use `contract` with certain indices only?
> 
> Yes! That's the 3rd argument. Pass a list or tuple of `Symbol`s.

Is this supposed to work like this? It seems that I can not get the expected result using `contract`:
```julia
julia> A = tensors(ϕ, 2); ein"ijk,ilk->jl"(A, conj(A))
4×4 Matrix{Float64}:
  1.0          -6.51726e-17  3.57857e-17  1.83402e-16
 -6.51726e-17   1.0          4.28191e-17  1.8131e-16
  3.57857e-17   4.28191e-17  1.0          7.81697e-17
  1.83402e-16   1.8131e-16   7.81697e-17  1.0

julia> A = tensors(ϕ, 2); Tenet.contract(A, conj(A), ((:i,:j,:k), (:i,:k,:l)))
2×4×2 Tensor{Float64, 3, Array{Float64, 3}}:
[:, :, 1] =
 9.86076e-32  0.0935065    0.0332468  0.873247
 1.0          2.60386e-31  0.0        0.0

[:, :, 2] =
 1.248e-31    0.841558   0.101834  0.0566071
 9.62965e-33  0.0649351  0.864919  0.0701461
```

**Comment by jofrevalles** [comment]
I updated the code to use now `contract` instead of `OMEinsum`. The discussion on the `labels` argument of the `contract` function for `Tensor`s is continued in [Tensors/#13](https://github.com/bsc-quantic/Tensors.jl/issues/13).

File: src/Quantum/MatrixProductState.jl [hunk 1]
```diff
@@ -185,105 +185,93 @@ function Base.rand(rng::Random.AbstractRNG, sampler::MPSSampler{Closed,T}) where
...
+    if dim isa Symbol
+        dim = findfirst(==(dim), labels(t))
+    end
+
+    data = view(parent(t), ntuple(i -> i == dim ? (1:end_ind) : Colon(), ndims(t))...)
+    new_labels = labels(t)
+
+    return Tensor(data, new_labels; t.meta...)
+end
+
```
Comment (mofeing) [inline src/Quantum/MatrixProductState.jl]: Don't use this function. Use `@view` on the tensor.

File: src/Quantum/MatrixProductState.jl [hunk 2]
```diff
@@ -185,105 +185,93 @@ function Base.rand(rng::Random.AbstractRNG, sampler::MPSSampler{Closed,T}) where
...
-        end
+        replace!(ψ, A => A_new)
+        replace!(ψ, B => B_new)
     end
 
-
-    # Create a new MPS object with the updated tensors
-    ψ = MatrixProductState(tensors_array)
-
     return_singular_values ? (ψ, singular_values) : ψ
```
Comment (mofeing) [inline src/Quantum/MatrixProductState.jl]: We can rewrite this part using the QR decomposition. It should be more efficient and easier to implement.

I can do that part.

File: src/Quantum/MatrixProductState.jl (hunk 1 above)
Comment (jofrevalles) [reply src/Quantum/MatrixProductState.jl]: I am using `view` in the code

File: src/Quantum/MatrixProductState.jl (hunk 2 above)
Comment (jofrevalles) [reply src/Quantum/MatrixProductState.jl]: Okay, but note that this is outdated code.

Comment (jofrevalles) [reply src/Quantum/MatrixProductState.jl]: You can see it here in this PR. The main function is:
```julia
function canonize!(ψ::TensorNetwork{MatrixProductState{Open}}, center::Union{Integer,UnitRange}; chi::Int = 0, return_singular_values::Bool=false)

    # Define the local helper function canonize_two_sites within canonize
    function canonize_two_sites(A::Tensor, B::Tensor, i::Int, center::Union{Integer,UnitRange}; chi::Int = 0, return_singular_values::Bool=false)
        left_edge, right_edge = extrema(center)

        A_alias = A.meta[:alias]
        B_alias = B.meta[:alias]

        C = contract(A, B)

        left_inds = (kv -> kv[2]).(filter(p -> p.first ≠ :r, collect(A_alias)))
        U, S, V = svd(C; left_inds=left_inds)

        # Truncate if desired bond dimension is provided
        if chi > 0 && chi < size(S, 1)
            U = view(U, labels(U)[end] => 1:chi)
            S = Tensor(Diagonal(view(S, labels(S)[1] => 1:chi,  labels(S)[2] => 1:chi)), labels(S))
            V = view(V, labels(V)[begin] => 1:chi)
        end

        # Reshape U and V and update tensors
        if i < left_edge # Move orthogonality center to the right
            A_new = replace(U, labels(U)[end]=> A_alias[:r])
            B_new = replace(V * S, labels(U)[end]=> B_alias[:l])
        elseif i > right_edge # Move orthogonality center to the left
            A_new = replace(U * S, labels(V)[begin]=> A_alias[:r])
            B_new = replace(V, labels(V)[begin]=> B_alias[:l])
        else # No need to update tensors
            A_new = A
            B_new = B
        end

        A_new = Tensor(parent(A_new), labels(A_new); alias = A_alias)
        B_new = Tensor(parent(B_new), labels(B_new); alias = B_alias)

        return return_singular_values ? (A_new, B_new, S) : (A_new, B_new)
    end

    N = length(ψ)
    return_singular_values ? singular_values = Vector{Vector{Float64}}(undef, N-1) : nothing

    # First sweep from left to right
    for i in 1:N-1
        A = tensors(ψ, i)
        B = tensors(ψ, i+1)

        A_new, B_new, S = return_singular_values ?
            canonize_two_sites(A, B, i, center; chi = chi, return_singular_values = true) :
            (canonize_two_sites(A, B, i, center; chi = chi)..., nothing) # Ignore S

        replace!(ψ, A => A_new)
        replace!(ψ, B => B_new)
    end

    # Second sweep from right to left to ensure canonical form
    for i in N-1:-1:1
        A = tensors(ψ, i)
        B = tensors(ψ, i+1)

        A_new, B_new, S = return_singular_values ?
            canonize_two_sites(A, B, i, center; chi = chi, return_singular_values = true) :
            (canonize_two_sites(A, B, i, center; chi = chi)..., nothing) # Ignore S

        # Normalize singular values with ./ sqrt(sum(parent(S).diag .^ 2))
        return_singular_values && (singular_values[i] = parent(S).diag ./ sqrt(sum(parent(S).diag .^ 2)))

        replace!(ψ, A => A_new)
        replace!(ψ, B => B_new)
    end

    return_singular_values ? (ψ, singular_values) : ψ
end
```

**Comment by mofeing** [comment]
Related code moved to [`Qrochet.jl`](https://github.com/bsc-quantic/Qrochet.jl).

**Pull request closed without merging**